pilot/pkg/xds/discovery.go

```diff
@@ -649,3 +655,11 @@ func (s *DiscoveryServer) ClientsOf(typeUrl string) []*Connection {
 
 	return pending
 }
+
+func (s *DiscoveryServer) WaitForRequestLimit(ctx context.Context) error {
+	// Give a bit of time for queue to clear out, but if not fail fast. Client will connect to another
+	// instance in best case, or retry with backoff.
+	wait, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	return s.requestRateLimit.Wait(wait)
```

Comment:
@howardjohn I think this can not handle 1k xds requests being handled concurrently. because of the duration of each request handling.

In other word, why not handle concurrent xds processing with buffered channel like what we have done for xds push.

Reply:
I discussed this a bit in the issue description. The problem we face in doing that is the request initialization has a number of expensive actions that are not done in the push. In particular, there is an API server call to create workload entry for VMs. If we have that locked by a semaphore, then even a small latency spike in api-server will cause xds to be blocked.

With this implementation it will not be. While it may not be ideal in every possible case it is very simple and should make every case behave better. In the extreme rare case that someone actually needs to serve over 100qps with a single instance, they can adjust it disable this

Reply:
This scenario is like kube-apiserver, it denies any request incoming if the total requests being precessed exceed the max inflight limit.  Which is far more safer for a server. Even with this, k8s introduced flow control recently https://kubernetes.io/docs/concepts/cluster-administration/flow-control/